protobufs/livekit_models.proto

```diff
@@ -302,6 +305,32 @@ message ChatMessage {
   bool generated = 6; // true if the chat message has been generated by an agent from a participant's audio transcription
 }
 
+message RpcRequest {
+  string id = 1;
+  string method = 2;
+  string payload = 3;
+  uint32 response_timeout_ms = 4;
+  uint32 version = 5;
+}
+
+message RpcAck {
+  string request_id = 1;
+}
+
+message RpcResponse {
+  string request_id = 1;
+  oneof value {
+    string payload = 2;
+    RpcError error = 3;
+  }
+}
+
+message RpcError {
+  string name = 1;
```

Comment:
do we want error codes instead of name ? for a lot of our RPCs, we return codes similar to HTTP in order to make it easier on users.

Reply:
the complexity here is that RpcError needs to support built-in errors _and_ user definable errors, but we could make them separate types if we wanted.

I think for our errors, number codes are probably fine (esp since there's a message field) but for user-defined errors string is much easier and doesn't have weird issues like reserved number space.

Reply:
I don't think we necessarily need a "reserved" space.. personal opinion is that numbers make it easier to reason with vs strings. 

psrpc usage is similar.. where psrpc is the protocol, but different services would return different errors, while still adhering to HTTP-like status codes.

Reply:
by reserved space I meant that we currently use name to mean a specific error, not a class of error.  if it's http-like then it's classes, which is different.

 these are the errors the [prototype](https://github.com/livekit/node-sdks/pull/276) has found cause to implement:

```
  CONNECTION_TIMEOUT (no ack)
  RESPONSE_TIMEOUT (no response)
  UNSUPPORTED_METHOD (no registered handler)
  RECIPIENT_DISCONNECTED (recipient dropped while responding)
  UNCAUGHT_ERROR (some random error was thrown in the handler - note that users may throw our error class instead if they want it passed through as-is)
  MALFORMED_RESPONSE (this is likely to go away when we switch from JSON to protobuf)
  PAYLOAD_TOO_LARGE (your request included too much data OR the response included too much data)
  ```
  
  these could be made HTTP-like as:
  
  ```
   503 connection timeout
   504 response timeout, recipient disconnected
   404 method not found
   500 response payload too large
   400 request payload too large
 ```

is that what you mean?

Reply:
we use an enum with a reserved space for metrics, guess we could do the same thing here?

Reply:
yeah i had it that way originally but I think it comes down to the developer ergonomics:


error handling (swifty pseudo-code)

```
try {
   performRpcRequest(....)
} catch (error: RpcError) {

   if (error.name == LKUnsupportedMethod) {
         log(error) // { name: "lk-unsupported-method", message: "this method has not been registered for RPC"}
      // handling for some livekit built in
   } else if (error.name == "My-Special-Error") {
            log(error) // { name: "My-Special-Error", message: "something terrible has happened"}
      // handling for your own error
   }
}
``` 

error throwing (swifty pseudo-code)

```
registerRpcHandler('my-method', { request in
   throw new RpcError("My-Special-Error", "something terrible has happened"); // forwarded
   throw new Error(....); // converted to LKUncaughtError
}
```

Those feel good to me. I feel like if you convert the names to numbers then it just feels weird throwing your own, because you're required to define a strict code system? And you need to ensure the message is quite explanatory or else the log lines will require consulting your error table to make sense of.

I don't feel too strongly about it and am happy to change if someone else does feel strongly.  Otherwise I think I'll keep it as I have it....

(btw HTTP seems appealing but since RPC doesn't map 1:1 to HTTP it makes it so you still really need error names, at which point you've turned the three-field error into a four-field error and I'm not sure it's particularly useful? I could be wrong)

Reply:
> personal opinion is that numbers make it easier to reason with vs strings.

I share @davidzhao's opinion, probably because an `errorCode` as number is quite common in general and also because we use numbers already in some other places. 
An advantage, I believe, is the possibility of grouping errors with numbers (that's the only similarity I'd see to HTTP statuses btw). 
But not a hill I'm going to die on.

Reply:
Ok after more reflection and reading opinions on the internet I think it's right to switch back to numeric codes as in the original proposal - one per error. And we'll populate the message as well, based on the code (for our own errors).  We'll reserve the space 1001-1999 for framework-level errors and suggest that developers use the space 101-999 for application-level errors. And I guess I'll store it as a unit32 (wish they had unit16 in proto3!)